src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -408,32 +372,13 @@ static BoundInterpolatedString createInterpolation(BoundExpression expression, I
         {
             Debug.Assert(binaryOperator.IsUnconvertedInterpolatedStringAddition);
 
-            var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
             var partsArrayBuilder = ArrayBuilder<ImmutableArray<BoundExpression>>.GetInstance();
 
-            BoundBinaryOperator? current = binaryOperator;
-
-            while (current != null)
+            binaryOperator.VisitBinaryOperatorInterpolatedString(partsArrayBuilder, (BoundUnconvertedInterpolatedString unconvertedInterpolatedString, ArrayBuilder<ImmutableArray<BoundExpression>> partsArrayBuilder) =>
             {
-                stack.Push(current);
-                partsArrayBuilder.Add(((BoundUnconvertedInterpolatedString)current.Right).Parts);
-
-                if (current.Left is BoundBinaryOperator next)
-                {
-                    current = next;
-                }
-                else
-                {
-                    partsArrayBuilder.Add(((BoundUnconvertedInterpolatedString)current.Left).Parts);
-                    current = null;
-                }
-            }
-
-            // Parts are added in right to left order, but lexical is left to right.
-            partsArrayBuilder.ReverseContents();
-
-            Debug.Assert(partsArrayBuilder.Count == stack.Count + 1);
-            Debug.Assert(partsArrayBuilder.Count >= 2);
+                partsArrayBuilder.Add(unconvertedInterpolatedString.Parts);
```

Comment:
>unconvertedInterpolatedString.Parts [](http://example.com/codeflow?start=38&length=35)

Would be useful to assert something about the parts? Validity, etc.

Reply:
I can't think of anything useful to assert here. The parts can have errors or be invalid in the context (such as having `awaits` within a `ref struct` builder type) and binding will issue errors as appropriate.